java/oauth-custom-challengehandler/src/main/java/com/esri/arcgisruntime/sample/oauthcustomchallengehandler/MainActivity.java

```diff
@@ -0,0 +1,298 @@
+                mOAuthConfiguration.getClientId(), mOAuthConfiguration.getRedirectUri());
+            webView.setWebViewClient(customWebViewClient);
+
+            webView.loadUrl(url);
+            // set the webview on the dialog
+            mOAuthLoginDialog.setContentView(webView);
+            // display the dialog
+            mOAuthLoginDialog.show();
+          }
+        });
+
+        try {
+          // Wait for the OAuth browser page to be shown and user to login and for the CustomWebViewClient
+          // to get the auth code and request an OAuthTokenCredential with it
+          sOAuthChallengeCountDownLatch.await();
+        } catch (InterruptedException ie) {
+        }
+        // Get the credential from the latch once it has been counted down
+        OAuthTokenCredential credential = sOAuthChallengeCountDownLatch.getOAuthTokenCredential();
+        if (credential != null) {
+          ret = new AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CONTINUE_WITH_CREDENTIAL,
+              credential);
+        }
+        // Reset the latch for the next challenge
+        sOAuthChallengeCountDownLatch = null;
+        // dismiss the dialog
+        mOAuthLoginDialog.dismiss();
+      }
+      return ret;
+    }
+  };
+
+  /**
+   * CustomWebViewClient overrides the shouldOverrideUrlLoading method to handle errors on the login page and to request
+   * OAuthTokenCredential after having received the auth code when the user provides the login credential/
+   */
+  private class CustomWebViewClient extends WebViewClient {
+
+    private String mPortal;
+
+    private String mClientId;
+
+    private String mRedirectUri;
+
+    CustomWebViewClient (String portal, String clientId, String redirectUri) {
+      mPortal = portal;
+      mClientId = clientId;
+      mRedirectUri = redirectUri;
+    }
+
+    @Override
+    public boolean shouldOverrideUrlLoading(WebView view, String url) {
+
+      Uri responseUri = Uri.parse(url);
+      String schema = responseUri.getScheme();
+      if (schema == null) {
+        throw new RuntimeException("Invalid schema could not be parsed: " + responseUri.toString());
+      }
+
+      String error = responseUri.getQueryParameter("error");
+      String errorDescription = responseUri.getQueryParameter("error_description");
+      final String authCode = responseUri.getQueryParameter("code");
+
+      // an error is returned from the portal OAuth login page when the user presses cancel button
+      // that case is handled here
+      if (error != null && !error.isEmpty()) {
+        if (error.equalsIgnoreCase("access_denied") && errorDescription.equalsIgnoreCase("The user denied your request.")) {
+          Toast.makeText(getApplicationContext(), "No credentials provided, failed to load secured resource", Toast.LENGTH_LONG).show();
+          mOAuthLoginDialog.dismiss();
+          return true;
```

Comment:
Shouldn't we call `finishOAuthChallenge` before returning?

Reply:
should be `return false`?

Reply:
> Shouldn't we call finishOAuthChallenge before returning? 

updated

> should be return false?    

if we return false, WebView will load the error page. For optics I decided against it.